src/function/gds/rec_joins.cpp

```diff
@@ -202,5 +202,22 @@ void RJAlgorithm::exec(processor::ExecutionContext* executionContext) {
     sharedState->mergeLocalTables();
 }
 
+bool SPAlgorithm::terminateFunc(GDSFrontier& frontier,
+    const processor::NodeOffsetMaskMap& maskMap) {
+    if (!maskMap.enabled()) {
+        return false;
+    }
+    for (auto& [tableID, mask] : maskMap.getMaskMap()) {
+        if (frontier.pinTableID(tableID)) {
+            for (auto i = 0u; i < mask->getMaxOffset(); ++i) {
+                if (mask->isMasked(i) && !frontier.isActive(i)) {
```

Comment:
This termination condition looks wrong to me. What if there are 2 output nodes vi and vj, so maskMap contains 2 nodes. But vi of them is reached in iteration 1, so will be active in frontier1 and vj in iteration 5, so in frontier 5. So there will never be a frontier in which both vi and vj are in the same frontier, although both of them have been reached. If this is a bug, we should add a test case for this.

Reply:
I don't think this is a bug. Or at least it's not a bug that can be captured by e2e test. It's just we won't trigger early termination.